Fix hash collisions issue

LKomaryanskiy opened the pull request:
In some cases, L3 offload does not work because the number of hash collisions becomes larger than zero while adding the rules to L3 table. To fix this issue, implemented 2 patches:
1) Set appropriate maximum number of collisions according to the formula from documentation:
FWLTHHEC.LTHHMC = (clk_freq[Hz] * Average_frame_size[bit] / Incoming_throughput[bps] - 4) / 3

2) As the maximum number of hash collisions can be exceeded (due to large number of L3 rules), implemented mechanism that is able to find more suitable hash equation and set appropriate value to FWLTHHC register to reduce collisions without dropping throughput.

LKomaryanskiy committed:
net: renesas: rswitch: set appropriate number of hash collision

Currently, the number of hash collisions is zero by default. It leads
to an issue when at least one collision makes L3 rule invalid due to
hardware algorithm that checks if FWLTHTLR.LTHLCN is less or equal
to FWLTHHEC.LTHHMC. The appropriate value should be calculated and
set based on the formula:
FWLTHHEC.LTHHMC =
(clk_freq[Hz] * Average_frame_size[bit] / Incoming_throughput[bps] - 4) / 3

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Volodymyr Babchuk <volodmyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: implement hash adjustment mechanism

In case of using a large number of stream filters the number
of hash collisions may be greater than available maximum. It
can be fixed by adjusting of hash equation register to reduce
number of hash collisions.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Volodymyr Babchuk <volodmyr_babchuk@epam.com>

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4299,6 +4299,54 @@ static void rswitch_fwd_init(struct rswitch_private *priv)
+static void rswitch_set_max_hash_collisions(struct rswitch_private *priv)
+{
+	/* 320Mhz */
+	const u64 clk_freq = 320 * 1000 * 1000;
```
How you obtained this frequency? Shouldn't you use `clk` subsystem for this?

LKomaryanskiy replied:
The configuration of frequency comes from ptp clock driver as a constant define for the register. So I decided to move my constant to more appropriate place.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4299,6 +4299,54 @@ static void rswitch_fwd_init(struct rswitch_private *priv)
+	/* 320Mhz */
+	const u64 clk_freq = 320 * 1000 * 1000;
+	/* Avarage frame size in bits */
+	const u64 average_frame_size = 64 * 8;
```
It is customary to hold such constans at the beginning of the file, as `#define`

LKomaryanskiy replied:
Good point, changed to define.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -4299,6 +4299,54 @@ static void rswitch_fwd_init(struct rswitch_private *priv)
+	 */
+	priv->max_collisions = (((clk_freq * average_frame_size) / (max_throughput)) - 4) / 3;
+
+	of_node_put(ports);
```
It is better to release `ports` right after use.  I.e. near line 4338

LKomaryanskiy replied:
Good point, moved.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2368,14 +2383,115 @@ static int rswitch_modify_l3fwd(struct l3_ipv4_fwd_param *param, bool delete)
+			do {
+				priv->hash_equation++;
+				if (priv->hash_equation > FWLTHHC_LTHHE_MASK)
+					return -E2BIG;
```
Not the best idea, because it has 100% chance that system will long uptime will stop adding the rules. Consider the following sequence of events:

 - I add very unlucky set of rules that moves `priv->hash_equation` to almost maximum value.
 - I remove some of the rules, so lower values of  `priv->hash_equation` might work as well
 - I add new rule that forces next iteration of hash formula update and system stucks at this check, albeit lower values of `priv->hash_equation`  may be valid

LKomaryanskiy replied:
Yes, you're absolutely right, I missed this scenario. I changed the logic. Currently, it's trying to find appropriate parameters from the beginning, and if we reach the original value the function gives up finding a new value.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2368,14 +2383,115 @@ static int rswitch_modify_l3fwd(struct l3_ipv4_fwd_param *param, bool delete)
+		if (rc == -EAGAIN) {
+			do {
+				priv->hash_equation++;
+				if (priv->hash_equation > FWLTHHC_LTHHE_MASK)
```
Is the `FWLTHHC_LTHHE_MASK` correct constant to be checked against? Should it be  `FWLTHHC_LTHHE_MAX`?

LKomaryanskiy replied:
Oh, yes, I wrote "MASK" because in this case, it has the same sense (because configuration values are placed at the lower bits of the register). Nevertheless, for sure, it's more clear to use "MAX" suffix in this case, so I changed it.

lorc reviewed (commented):
Did you made some measurements on how long it takes to restore all the tables? Especially if they are filled by, say, 50%?

LKomaryanskiy reviewed (commented):
@lorc thank you for your close review, it really helped to avoid further issues.

> Did you made some measurements on how long it takes to restore all the tables? Especially if they are filled by, say, 50%?

I rechecked this and while making measurements found the issue in my code:
I was waiting wrong value in `FWLTHTIM` register (firstly it was `1`, but the correct one is `BIT(1)`). Because of this mistake, resetting L3 table took 1 second which was horrible. But after fixing this, I made accurate measurements and here are the results:

> ===== resetting complete time = 1108020 usec (1 msec)
> ===== recovering complete time = 2671440 usec (2 msec)
> ===== recovered 486 entries

One iteration with resetting and recovering almost 500 entries takes around 3.5 msec. So, even in the worst scenario, searching through all parameters (from 2 to 0x1FF) will take 1.782 sec. Also, please note a few aspects:
1) Currently, enabled using 512 entries, so 486 - is close to maximum value, not even 50%;
2) This is the worst scenario when we recover all entries on every iteration, but in most cases recovering will stop in the middle, exactly when the number of collisions exceeds the available value.
3) The last but I think the most important: changing hash equation parameters is a very rare situation, because the first patch in this PR configures the appropriate value to maximum hash collision register (was zero, the current value is 13). To test this exceptional scenario, I forced this number to `0`, but even when I increased it to `1`, I got only 3 invocations of "reset-recovery" part with 480+ entries. So, most likely, with an appropriate value in hash maximum collision number won't ever happen, but we definitely have to handle this, even if it's very rare situation.

LKomaryanskiy commented:
Updates in the last PR version:
1) Changed algorithm for searching the appropriate parameters: now, when maximum value is exceeded, it starts searching from the beginning. In this case, the board with large uptime shouldn't have issue, when we won't be able to add new entries;
2) Fixed long reset due to wrong waiting criteria;
3) Provided accurate time measurements of "reset-recovery" part: https://github.com/xen-troops/linux/pull/159#pullrequestreview-1621505589
4) Small cosmetic fixes (some constants made as define etc)

lorc reviewed (approved):
Thank you for the measurements. I agree, that 3.5ms for table refresh is adequate timespan.

`Reviewed-by: Volodymyr Babchuk <volodmyr_babchuk@epam.com>`

LKomaryanskiy commented:
@lorc thank you. Added reviewed-by tags.
